### PR TITLE
CircleCI Enhancements

### LINKED ISSUE #505: CircleCI Enchancements
## Is your feature request related to a problem? Please describe.
- Enhance CircleCI configuration for independent build and test jobs.
- Support for workflows like smoke \ regression \ performance along with scheduling on specific branches
- Better error reporting \ feedback on test failures
- Optimize build time by eliminating creation of a new docker image on every build trigger
- Support for executing CircleCI jobs on developer's local machine

## Describe the solution
1. Create a docker image tailored for torchserve builds, make it available in public dockerhub registry
2. In CircleCI's config.yaml
    - Separate torchseve\modelarchiver build and tests as independent jobs
        - Use the docker image(from Step1) as an executor for all the jobs
        - Add new jobs like - torchserve build, python tests, newman tests, benchmark, modelarchiver build and tests
    - Each job to store respective build\test artifacts for better reporting
    - Create and use workflows along with triggers\schedule\branches to orchestrate jobs
3. Create a script using "circleci cli" to help developers execute a specific job \ all jobs from a workflow on their local machine

## Describe alternatives solution
\-

### PR DESCRIPTION
**Pull request by prashantsail**
## Description
Enhancements to CircleCI -
1. Created a docker image tailored for torchserve builds (_.circleci/images/Dockerfile_)
    The images are built locally and are currently available on [AWS ECR](https://console.aws.amazon.com/ecr/repositories/torchserve-build/?region=us-east-1)
2. In CircleCI's config.yaml
    - New Jobs added for torchserve build, python tests, newman api tests, benchmark, modelarchiver build and tests
    - All jobs use the docker image(from Step1) as an executor
    - All jobs are orchestrated using their respective shell scripts
       (Avoiding inline commands in config.yaml helps, with easier maintenance)
    - All jobs store respective build\test artifacts for better reporting
    - Added following workflows sanity(on every commit), regression and performance(scheduled only on master branch)
3. Created a python script(_run_circleci_tests.py_) which uses "circleci cli" to help developers execute a specific job \ all jobs from a workflow on their local machine
4. All details are available in _.circleci/Readme.md_

Fixes #505

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Feature/Issue validation/testing

[Worklfow showing all jobs being triggered](https://app.circleci.com/pipelines/github/prashantsail/serve/75/workflows/9181fb17-05d1-4e93-99c7-182c1a1daa06)
Screenshot 2020-07-16 at 00 32 35

> Note:
> 1. Model archiver build is failing due to pyilnt  
> 2.  API test and benchmark are failing due to limitiation of running on a medium resource(2CPU, 4GB RAM) container available for free plan

## Checklist:

- [x] Have you added tests that prove your fix is effective or that this feature works?
- [x] New and existing unit tests pass locally with these changes?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

### CODE REVIEW

**Comment by prashantsail** [comment]
ToDo
- [ ] Update PR's Description section with results and logs
- [ ] Move the docker image from my personal registry to pytorch\torchseve registry
- [ ] Newman api tests setup and success
- [ ] Add support for snapshot testcase from regression suite
- [ ] Circle smoke workflow at par with _torchserve_sanity.sh_
- [ ] Add support for Apache Benchmark as part of weekly workflow

**Comment by seemethere** [comment]
I can turn CircleCI commit status reporting back on if that'll help with the development of this PR, and if we can get the @sagemaker-neo-ci-bot to stop spamming my emails with notifications.

File: .circleci/README.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+- _-j, --job job_name_  
+If specified, executes only the specified job name (along with the required parents).  
+If not specified, all jobs in the workflow are executed sequentially.  
+
+- _-e, --executor executor_name_  
+If specified, job is executed only on the specified executor(docker image).  
+If not specified, job is executed on all the available executors.  
+
+```bash
```
Comment (dhanainme) [inline .circleci/README.md:60]: Can you move this above. As this is going to be the most frequently used commands.

File: .circleci/README.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# Pytorch Serve CircleCI build
+TorchServe uses CircleCI for builds. This folder contains the config and scripts that are needed for CircleCI.
+
+## config.yml
+_config.yml_ contains TorchServe's build logic which will be used by CircleCI.
+
+## Workflows and Jobs
+Currently, following _workflows_ are available -
```
Comment (dhanainme) [inline .circleci/README.md:8]: Maintain existing names - torch-serve-build-gpu / torch-serve-build-cpu / torch-serve-regression-test-cpu / torch-serve-regression-test-gpu / torch-serve-regression-performance-test-cpu / torch-serve-regression-performance-test-gpu

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,74 +1,298 @@
 version: 2.1
 
+
+executors:
+  py36:
+    docker:
+      - image: shivamshriwas/torchserve:py36
```
Comment (dhanainme) [inline .circleci/config.yml:7]: Move this to ECR & Use Credentials as env variables - https://github.com/pytorch/pytorch/blob/master/.circleci/config.yml

File: .circleci/config.yml [hunk 4]
```diff
@@ -1,74 +1,298 @@
...
+    environment:
+      _JAVA_OPTIONS: "-Xmx2048m"
+  conda38:
```
Comment (dhanainme) [inline .circleci/config.yml:10]: Make the env names consistent with https://github.com/pytorch/pytorch/blob/master/.circleci/config.yml

File: .circleci/config.yml [hunk 5]
```diff
@@ -1,74 +1,298 @@
...
+  pyenv37:
+    docker:
+      - image: shivamshriwas/torchserve:pyenv37
+    environment:
+      _JAVA_OPTIONS: "-Xmx2048m"
+  venv36:
+    docker:
+      - image: shivamshriwas/torchserve:venv36
+    environment:
+      _JAVA_OPTIONS: "-Xmx2048m"
```
Comment (dhanainme) [inline .circleci/config.yml:24]: _JAVA_OPTIONS may be redundant.

File: .circleci/scripts/linux_test_benchmark.sh [hunk 6]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
```
Comment (dhanainme) [inline .circleci/scripts/linux_test_benchmark.sh:1]: Can you please create an issue to track the removal of redundant scripts.

File: .circleci/images/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        ca-certificates\
+        curl \
+        git \
+        openjdk-11-jdk \
+    && curl -sL https://deb.nodesource.com/setup_14.x | bash - \
+    && apt-get update \
+    && apt-get install --no-install-recommends -y \
+        nodejs
+
+FROM compile-image as conda38-image
```
Comment (dhanainme) [inline .circleci/images/Dockerfile:20]: Can you document the build steps for this Docker file

File: .circleci/images/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        psutil \
+        pylint==2.4.4 \
+        pytest \
+        pytest-cov \
+        pytest-mock \
+        sentencepiece==0.1.91 \
+        torch \
+        torchtext \
+        torchvision \
+        transformers\
```
Comment (dhanainme) [inline .circleci/images/Dockerfile:99]: Can you pin versions

File: CODE_OF_CONDUCT.md [hunk 9]
```diff
@@ -55,7 +55,7 @@ a project may be further defined and clarified by project maintainers.
 ## Enforcement
 
 Instances of abusive, harassing, or otherwise unacceptable behavior may be
-reported by contacting the project team at <conduct@pytorch.org>. All
+reported by contacting the project team at \<conduct@pytorch.org\>. All
 complaints will be reviewed and investigated and will result in a response that
```
Comment (dhanainme) [inline CODE_OF_CONDUCT.md:59]: Pls confirm that this is not an issue.

File: run_circleci_tests.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        PROCESSED_CONFIG = yaml.safe_load(fstream)
+    except yaml.YAMLError as err:
+        print(err)
+
+# All executors available in the config file
+available_executors = list(CCI_CONFIG['executors'])
+
+# All jobs available under the specified workflow
+jobs_in_workflow = PROCESSED_CONFIG['workflows'][workflow]['jobs']
+
```
Comment (dhanainme) [inline run_circleci_tests.py:69]: Pls move then to a function.

File: .circleci/README.md (hunk 1 above)
Comment (prashantsail) [reply .circleci/README.md:27]: Done

File: .circleci/README.md (hunk 2 above)
Comment (prashantsail) [reply .circleci/README.md]: Workflow names revised to sanity, regression, performance

File: .circleci/config.yml (hunk 3 above)
Comment (prashantsail) [reply .circleci/config.yml]: Moved docker images from dockerhub to AWS ECR

File: .circleci/config.yml (hunk 4 above)
Comment (prashantsail) [reply .circleci/config.yml]: executor names like torchserve
example: ubuntu18-pythn36-cpu-docker

File: .circleci/images/Dockerfile (hunk 7 above)
Comment (prashantsail) [reply .circleci/images/Dockerfile:22]: Added comments in the docker file and also updated Readme for image creation

File: .circleci/images/Dockerfile (hunk 8 above)
Comment (prashantsail) [reply .circleci/images/Dockerfile]: Added versions for all torch packages

File: CODE_OF_CONDUCT.md (hunk 9 above)
Comment (prashantsail) [reply CODE_OF_CONDUCT.md]: Reverted this unintended change

File: .circleci/scripts/linux_test_benchmark.sh (hunk 6 above)
Comment (prashantsail) [reply .circleci/scripts/linux_test_benchmark.sh:1]: I will create an issue and all relevant details as soon as this PR is merged

**Comment by prashantsail** [comment]
GPU jobs \ scripts and clean up of redundant scripts will be take up in a separate PR

**Review by seemethere — APPROVED** [review]

**Review by dhanainme — APPROVED** [review]
Have pushed 

630887156731.dkr.ecr.us-east-1.amazonaws.com/torchserve-build:ubuntu18-conda38-cpu
630887156731.dkr.ecr.us-east-1.amazonaws.com/torchserve-build:ubuntu18-pythn36-cpu
630887156731.dkr.ecr.us-east-1.amazonaws.com/torchserve-build:ubuntu18-pyenv37-cpu
630887156731.dkr.ecr.us-east-1.amazonaws.com/torchserve-build:ubuntu18-venv36-cpu

to ECR.

Looks good to me.

**Review by dhanainme — APPROVED** [review]

**Pull request merged**